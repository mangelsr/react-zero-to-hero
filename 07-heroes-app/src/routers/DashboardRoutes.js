import React from 'react';
import { Redirect, Route, Switch } from 'react-router-dom';

import { DcScreen } from '../components/dc/DcScreen';
import { MarvelScreen } from '../components/marvel/MarvelScreen';
import { SearchScreen } from '../components/search/SearchScreen';
import { HeroScreen } from '../components/ui/heroes/HeroScreen';
import { Navbar } from '../components/ui/Navbar';


export const DashboardRoutes = () => {
    return (
        <>
            <Navbar />
            <div className="container mt-3">
                <Switch>
                    <Route exact path="/dc" component={DcScreen} />
                    <Route exact path="/marvel" component={MarvelScreen} />
                    <Route exact path="/hero/:heroId" component={HeroScreen} />
                    <Route exact path="/search" component={SearchScreen} />
                    <Redirect to="/dc" />
                </Switch>
            </div>
        </>
    )
}
