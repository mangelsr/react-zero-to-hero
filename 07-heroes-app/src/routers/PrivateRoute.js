import React from 'react'
import { Redirect, Route } from 'react-router-dom';
import PropTypes from 'prop-types';


export const PrivateRoute = ({ 
    isAuth,
    component: Component,
    ...rest
}) => {

    localStorage.setItem('lastpath', rest.location.pathname);

    return (
        <Route {...rest}
            component={props => isAuth
                ? <Component {...props} /> 
                : <Redirect to="/login" />}
            >            
        </Route>
    );
}

PrivateRoute.propTypes = {
    isAuth: PropTypes.bool.isRequired,
    component: PropTypes.func.isRequired
};