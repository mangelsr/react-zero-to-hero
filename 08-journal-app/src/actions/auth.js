import { firebase, googleAuthProvider } from "../firebase/firebaseConfig";
import Swal from "sweetalert2";

import { types } from "../types/types";
import { uiEndLoading, uiStartLoading } from "./ui";
import { clearNotes } from "./notes";

export const startLoginEmailPassword = (email, password) => {
    return async (dispatch) => {
        try {
            dispatch(uiStartLoading());
            const { user } = await firebase.auth().signInWithEmailAndPassword(email, password);
            dispatch(login(user.uid, user.displayName));
        } catch (e) {
            console.log(e);
            if (e.message) {
                Swal.fire('Error', e.message, 'error');
            }
        } finally {
            dispatch(uiEndLoading());
        }
    };
}

export const startGoogleLogin = () => {
    return async (dispatch) => {
        try {
            dispatch(uiStartLoading());
            const { user } = await firebase.auth().signInWithPopup(googleAuthProvider);
            dispatch(login(user.uid, user.displayName))
        } catch (e) {
            console.log(e);
            if (e.message) {
                Swal.fire('Error', e.message, 'error');
            }
        } finally {
            dispatch(uiEndLoading());
        }
    };
};


export const startRegister = (email, password, name) => {
    return async (dispatch) => {
        try {
            dispatch(uiStartLoading());
            const { user } = await firebase.auth().createUserWithEmailAndPassword(email, password);
            await user.updateProfile({displayName: name});
            dispatch(login(user.uid, user.displayName));
        } catch (e) {
            console.log(e);
            if (e.message) {
                Swal.fire('Error', e.message, 'error');
            }
        } finally {
            dispatch(uiEndLoading());
        }
    };
};

export const login = (uid, displayName) => ({
    type: types.login,
    payload: {
        uid,
        displayName
    }
})

export const startLogout = () => {
    return async (dispatch) => {
        try {
            await firebase.auth().signOut();
            dispatch(logout());
            dispatch(clearNotes());
        } catch (e) {
            console.log(e);
            if (e.message) {
                Swal.fire('Error', e.message, 'error');
            }
        }
    };
};

export const logout = () => ({
    type: types.logout
});