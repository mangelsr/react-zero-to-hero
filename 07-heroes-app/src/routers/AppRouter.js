import React, { useContext } from 'react';
import {
    BrowserRouter as Router,
    Switch
} from "react-router-dom";

import { AuthContext } from '../auth/AuthContext';
import { LoginScreen } from '../components/login/LoginScreen';
import { DashboardRoutes } from './DashboardRoutes';
import { PrivateRoute } from './PrivateRoute';
import { PublicRoute } from './PublicRoute';


export const AppRouter = () => {

    const { user:{logged} } = useContext(AuthContext);

    return (
        <div>
            <Router>
            <div>
                <Switch>
                    <PublicRoute isAuth={logged} exact path="/login" component={LoginScreen} />
                    <PrivateRoute isAuth={logged} path="/" component={DashboardRoutes} />
                </Switch>
            </div>
            </Router>
        </div>
    )
}
