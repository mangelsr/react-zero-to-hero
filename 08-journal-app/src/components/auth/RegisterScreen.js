import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import validator from "validator";

import { startRegister } from '../../actions/auth';
import { removeError, setError } from '../../actions/ui';
import { useForm } from '../../hooks/useForm';


export const RegisterScreen = () => {

    const dispatch = useDispatch();
    const { loading, msgError } = useSelector(state => state.ui);

    const [ formValues, handleInputChange ] = useForm({
        name: 'Angel',
        email: 'angel@gmail.com',
        password: '123456',
        password2: '123456',
    });

    const { name, email, password, password2 } = formValues;

    const handleRegister = e => {
        e.preventDefault();
        if (isFormValid()) {
            dispatch(startRegister(email, password, name));
        }
    };

    const isFormValid = () => {
        if (name.trim().length === 0) {
            dispatch(setError('Name is required'));
            return false;
        } else if (!validator.isEmail(email)) {
            dispatch(setError('Email is not valid'));
            return false;
        } else if (password !== password2 || password.length < 5) {
            dispatch(setError('Password should match and be at least 6 characers'));
            return false;
        }
        dispatch(removeError());
        return true;
    };

    return (
        <div className="animate__animated animate__fadeIn">
            <h3 className="auth__title">Register</h3>
            <form onSubmit ={handleRegister}>

                { msgError && 
                    <div className="auth__alert-error">
                        { msgError }
                    </div>
                }
                
                <input onChange={handleInputChange} value={name} className="auth__input" type="text" placeholder="Name" name="name" autoComplete="off"/>
                <input onChange={handleInputChange} value={email} className="auth__input" type="text" placeholder="Email" name="email" autoComplete="off"/>
                <input onChange={handleInputChange} value={password} className="auth__input" type="password" placeholder="Password" name="password"/>
                <input onChange={handleInputChange} value={password2} className="auth__input" type="password" placeholder="Password confirmation" name="password2"/>
                
                <button disabled={loading} className="btn btn-primary btn-block mb-5" type="submit">Register</button>

                <Link className="link" to="/auth/login">
                    Already registered?
                </Link>
            </form>
        </div>
    );
}
