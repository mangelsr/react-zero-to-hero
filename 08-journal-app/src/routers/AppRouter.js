import React, { useEffect, useState } from 'react';
import { BrowserRouter as Router, Switch, Redirect } from 'react-router-dom';
import { useDispatch } from 'react-redux';

import { AuthRouter } from './AuthRouter';
import { JournalScreen } from '../components/journal/JournalScreen';
import { firebase } from "../firebase/firebaseConfig";
import { login } from '../actions/auth';
import { PublicRoute } from './PublicRoute';
import { PrivateRoute } from './PrivateRoute';
import { startLoadingNotes } from '../actions/notes';


export const AppRouter = () => {

    const dispatch = useDispatch();
    const [checking, setChecking] = useState(true);
    const [isLoggedIn, setIsLoggedIn] = useState(false);

    useEffect(() => {
        firebase.auth().onAuthStateChanged(async (user) => {
            if (user?.uid) {
                dispatch(login(user.uid, user.displayName));
                setIsLoggedIn(true);
                dispatch(startLoadingNotes(user.uid)) ;
            } else {
                setIsLoggedIn(false);
            }
            setChecking(false);
        });
    }, [dispatch, setChecking, setIsLoggedIn]);

    if (checking) {
        return <h1>Wait...</h1>;
    } else {
        return (
            <Router>
                <Switch>
                    <PublicRoute isAuth={isLoggedIn} path="/auth" component={AuthRouter} />
                    <PrivateRoute exact isAuth={isLoggedIn} path="/" component={JournalScreen} />
                    <Redirect to="/" />
                </Switch>
            </Router>
        );
    }
}
