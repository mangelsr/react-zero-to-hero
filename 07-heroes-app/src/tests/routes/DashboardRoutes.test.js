import '@testing-library/jest-dom';
import React from 'react';
import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';

import { AuthContext } from '../../auth/AuthContext';
import { DashboardRoutes } from '../../routers/DashboardRoutes';


describe('Tests over <DashboardRoutes /> component', () => {

    const contextValue = {
        dispatch: jest.fn(),
        user: {
            logged: true,
            name: 'Miguel'
        }
    };
    
    test('should rendered correctly', () => {
        const wrapper = mount(
            <AuthContext.Provider value={contextValue}>
                <MemoryRouter>
                    <DashboardRoutes/>
                </MemoryRouter>
            </AuthContext.Provider>
        );

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('span.text-info').text().trim()).toBe(contextValue.user.name);
    });
    
});