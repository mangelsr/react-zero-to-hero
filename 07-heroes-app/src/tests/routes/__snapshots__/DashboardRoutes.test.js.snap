// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests over <DashboardRoutes /> component should rendered correctly 1`] = `
Array [
  <nav
    className="navbar navbar-expand-sm navbar-dark bg-dark"
  >
    <a
      className="navbar-brand"
      href="/"
      onClick={[Function]}
    >
      Asociaciones
    </a>
    <div
      className="navbar-collapse"
    >
      <div
        className="navbar-nav"
      >
        <a
          aria-current="page"
          className="nav-item nav-link active"
          href="/dc"
          onClick={[Function]}
          style={Object {}}
        >
          DC
        </a>
        <a
          aria-current={null}
          className="nav-item nav-link"
          href="/marvel"
          onClick={[Function]}
        >
          Marvel
        </a>
        <a
          aria-current={null}
          className="nav-item nav-link"
          href="/search"
          onClick={[Function]}
        >
          Search
        </a>
      </div>
    </div>
    <div
      className="navbar-collapse collapse w-100 order-3 dual-collapse2"
    >
      <ul
        className="navbar-nav ml-auto"
      >
        <span
          className="nav-item nav-link text-info"
        >
          Miguel
        </span>
        <button
          className="nav-item nav-link btn"
          onClick={[Function]}
        >
          Logout
        </button>
      </ul>
    </div>
  </nav>,
  <div
    className="container mt-3"
  >
    <div>
      <h1>
        DC
      </h1>
      <hr />
      <div
        className="card-columns animate__animated animate__fadeIn"
      >
        <div
          className="card ms-3"
          style={
            Object {
              "maxWidth": 540,
            }
          }
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-md-4"
            >
              <img
                alt="Batman"
                className="card-img"
                src="./assets/dc-batman.jpg"
              />
            </div>
            <div
              className="col-md-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Batman
                </h5>
                <p
                  className="card-text"
                >
                  Bruce Wayne
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  >
                    Detective Comics #27
                  </small>
                </p>
                <a
                  className="btn btn-outline-primary"
                  href="/hero/dc-batman"
                  onClick={[Function]}
                >
                  Mas...
                </a>
              </div>
            </div>
          </div>
        </div>
        <div
          className="card ms-3"
          style={
            Object {
              "maxWidth": 540,
            }
          }
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-md-4"
            >
              <img
                alt="Superman"
                className="card-img"
                src="./assets/dc-superman.jpg"
              />
            </div>
            <div
              className="col-md-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Superman
                </h5>
                <p
                  className="card-text"
                >
                  Kal-El
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  >
                    Action Comics #1
                  </small>
                </p>
                <a
                  className="btn btn-outline-primary"
                  href="/hero/dc-superman"
                  onClick={[Function]}
                >
                  Mas...
                </a>
              </div>
            </div>
          </div>
        </div>
        <div
          className="card ms-3"
          style={
            Object {
              "maxWidth": 540,
            }
          }
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-md-4"
            >
              <img
                alt="Flash"
                className="card-img"
                src="./assets/dc-flash.jpg"
              />
            </div>
            <div
              className="col-md-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Flash
                </h5>
                <p
                  className="card-text"
                >
                  Jay Garrick
                </p>
                <p
                  className="card-text"
                >
                  Jay Garrick, Barry Allen, Wally West, Bart Allen
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  >
                    Flash Comics #1
                  </small>
                </p>
                <a
                  className="btn btn-outline-primary"
                  href="/hero/dc-flash"
                  onClick={[Function]}
                >
                  Mas...
                </a>
              </div>
            </div>
          </div>
        </div>
        <div
          className="card ms-3"
          style={
            Object {
              "maxWidth": 540,
            }
          }
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-md-4"
            >
              <img
                alt="Green Lantern"
                className="card-img"
                src="./assets/dc-green.jpg"
              />
            </div>
            <div
              className="col-md-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Green Lantern
                </h5>
                <p
                  className="card-text"
                >
                  Alan Scott
                </p>
                <p
                  className="card-text"
                >
                  Alan Scott, Hal Jordan, Guy Gardner, John Stewart, Kyle Raynor, Jade, Sinestro, Simon Baz
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  >
                    All-American Comics #16
                  </small>
                </p>
                <a
                  className="btn btn-outline-primary"
                  href="/hero/dc-green"
                  onClick={[Function]}
                >
                  Mas...
                </a>
              </div>
            </div>
          </div>
        </div>
        <div
          className="card ms-3"
          style={
            Object {
              "maxWidth": 540,
            }
          }
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-md-4"
            >
              <img
                alt="Green Arrow"
                className="card-img"
                src="./assets/dc-arrow.jpg"
              />
            </div>
            <div
              className="col-md-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Green Arrow
                </h5>
                <p
                  className="card-text"
                >
                  Oliver Queen
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  >
                    More Fun Comics #73
                  </small>
                </p>
                <a
                  className="btn btn-outline-primary"
                  href="/hero/dc-arrow"
                  onClick={[Function]}
                >
                  Mas...
                </a>
              </div>
            </div>
          </div>
        </div>
        <div
          className="card ms-3"
          style={
            Object {
              "maxWidth": 540,
            }
          }
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-md-4"
            >
              <img
                alt="Wonder Woman"
                className="card-img"
                src="./assets/dc-wonder.jpg"
              />
            </div>
            <div
              className="col-md-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Wonder Woman
                </h5>
                <p
                  className="card-text"
                >
                  Princess Diana
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  >
                    All Star Comics #8
                  </small>
                </p>
                <a
                  className="btn btn-outline-primary"
                  href="/hero/dc-wonder"
                  onClick={[Function]}
                >
                  Mas...
                </a>
              </div>
            </div>
          </div>
        </div>
        <div
          className="card ms-3"
          style={
            Object {
              "maxWidth": 540,
            }
          }
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-md-4"
            >
              <img
                alt="Martian Manhunter"
                className="card-img"
                src="./assets/dc-martian.jpg"
              />
            </div>
            <div
              className="col-md-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Martian Manhunter
                </h5>
                <p
                  className="card-text"
                >
                  J'onn J'onzz
                </p>
                <p
                  className="card-text"
                >
                  Martian Manhunter
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  >
                    Detective Comics #225
                  </small>
                </p>
                <a
                  className="btn btn-outline-primary"
                  href="/hero/dc-martian"
                  onClick={[Function]}
                >
                  Mas...
                </a>
              </div>
            </div>
          </div>
        </div>
        <div
          className="card ms-3"
          style={
            Object {
              "maxWidth": 540,
            }
          }
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-md-4"
            >
              <img
                alt="Robin/Nightwing"
                className="card-img"
                src="./assets/dc-robin.jpg"
              />
            </div>
            <div
              className="col-md-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Robin/Nightwing
                </h5>
                <p
                  className="card-text"
                >
                  Dick Grayson
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  >
                    Detective Comics #38
                  </small>
                </p>
                <a
                  className="btn btn-outline-primary"
                  href="/hero/dc-robin"
                  onClick={[Function]}
                >
                  Mas...
                </a>
              </div>
            </div>
          </div>
        </div>
        <div
          className="card ms-3"
          style={
            Object {
              "maxWidth": 540,
            }
          }
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-md-4"
            >
              <img
                alt="Blue Beetle"
                className="card-img"
                src="./assets/dc-blue.jpg"
              />
            </div>
            <div
              className="col-md-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Blue Beetle
                </h5>
                <p
                  className="card-text"
                >
                  Dan Garret
                </p>
                <p
                  className="card-text"
                >
                  Dan Garret, Ted Kord, Jaime Reyes
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  >
                    Mystery Men Comics #1
                  </small>
                </p>
                <a
                  className="btn btn-outline-primary"
                  href="/hero/dc-blue"
                  onClick={[Function]}
                >
                  Mas...
                </a>
              </div>
            </div>
          </div>
        </div>
        <div
          className="card ms-3"
          style={
            Object {
              "maxWidth": 540,
            }
          }
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-md-4"
            >
              <img
                alt="Black Canary"
                className="card-img"
                src="./assets/dc-black.jpg"
              />
            </div>
            <div
              className="col-md-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Black Canary
                </h5>
                <p
                  className="card-text"
                >
                  Dinah Drake
                </p>
                <p
                  className="card-text"
                >
                  Dinah Drake, Dinah Lance
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  >
                    Flash Comics #86
                  </small>
                </p>
                <a
                  className="btn btn-outline-primary"
                  href="/hero/dc-black"
                  onClick={[Function]}
                >
                  Mas...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;
